src/colorizer/CanvasWithOverlay.ts

```diff
@@ -109,325 +107,142 @@ export default class CanvasWithOverlay extends ColorizeCanvas {
+        colorMapRangeMin: this.colorMapRangeMin,
+        colorMapRangeMax: this.colorMapRangeMax,
+      },
+      legendStyle: this.legendStyle,
+    };
+    return getFooterRenderer(this.ctx, params, this.footerStyle);
   }
 
   /**
-   * Render the overlay to the canvas.
+   * Render the viewport canvas with overlay elements composited on top of it.
    */
   render(): void {
-    const devicePixelRatio = this.getPixelRatio();
-    this.canvas.width = this.canvasWidth * devicePixelRatio;
-    this.canvas.height = this.canvasHeight * devicePixelRatio;
+    // Expand size by header + footer, if rendering:
+    const headerRenderer = this.getHeaderRenderer(this.isHeaderVisibleOnExport && this.isExporting);
+    const footerRenderer = this.getFooterRenderer(this.isFooterVisibleOnExport && this.isExporting);
+    this.headerSize = headerRenderer.sizePx;
+    this.footerSize = footerRenderer.sizePx;
+
+    const devicePixelRatio = getPixelRatio();
+    this.canvas.width = Math.round(this.canvasSize.x * devicePixelRatio);
+    this.canvas.height = Math.round((this.canvasSize.y + this.headerSize.y + this.footerSize.y) * devicePixelRatio);
+
     //Clear canvas
     this.ctx.clearRect(0, 0, this.canvas.width, this.canvas.height);
 
     // Because CanvasWithOverlay is a child of ColorizeCanvas, this renders the base
     // colorized viewport image. It is then composited into the CanvasWithOverlay's canvas.
     super.render();
     this.ctx.imageSmoothingEnabled = false;
-    this.ctx.drawImage(super.domElement, 0, 0);
-
-    // Get dimensions + render methods for the elements, but don't render yet so we can draw the background
-    // behind them.
-    const origin = this.backgroundOptions.marginPx.clone().add(this.backgroundOptions.paddingPx);
-    const { sizePx: scaleBarDimensions, render: renderScaleBar } = this.getScaleBarRenderer(origin);
-    origin.y += scaleBarDimensions.y;
-    const { sizePx: timestampDimensions, render: renderTimestamp } = this.getTimestampRenderer(origin);
-
-    // If both elements are invisible, don't render the background.
-    if (scaleBarDimensions.equals(new Vector2(0, 0)) && timestampDimensions.equals(new Vector2(0, 0))) {
-      return;
-    }
+    this.ctx.drawImage(super.domElement, 0, Math.round(this.headerSize.y * devicePixelRatio));
 
-    // Draw background box behind the elements
-    const contentSize = new Vector2(
-      Math.max(scaleBarDimensions.x, timestampDimensions.x),
-      scaleBarDimensions.y + timestampDimensions.y
-    );
-    const boxSize = contentSize.clone().add(this.backgroundOptions.paddingPx.clone().multiplyScalar(2.0));
-    this.renderBackground(boxSize, this.backgroundOptions);
-
-    // Draw elements over the background
-    renderScaleBar();
-    renderTimestamp();
+    this.ctx.scale(devicePixelRatio, devicePixelRatio);
+
+    headerRenderer.render(new Vector2(0, 0));
+    footerRenderer.render(new Vector2(0, this.canvasSize.y + this.headerSize.y));
```

Comment:
All render methods that take in a Vector2 `origin` param position their rendered elements so the top left corner is at the origin.